sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -1257,306 +1263,234 @@ __ESIMD_API std::enable_if_t<std::is_floating_point<T>::value, T> asin(T src0) {
+ESIMD_INLINE __ESIMD_NS::simd<T, N> fmod(__ESIMD_NS::simd<T, N> y,
+                                         __ESIMD_NS::simd<T, N> x) {
+  static_assert(std::is_floating_point<T>::value,
+                "Floating point argument type is expected.");
+
+  __ESIMD_NS::simd<T, N> abs_x = __ESIMD_NS::abs(x);
+  __ESIMD_NS::simd<T, N> abs_y = __ESIMD_NS::abs(y);
 
-  __ESIMD_NS::simd<float, N> reminder =
-      abs_y - abs_x * __ESIMD_NS::trunc<float>(abs_y / abs_x);
+  using BitMaskType = std::conditional_t<sizeof(T) == 4, int32_t, int64_t>;
+  constexpr BitMaskType signBitMask =
+      sizeof(T) == 4 ? 0x80000000 : 0x8000000000000000;
 
-  abs_x.merge(0.0f, reminder >= 0);
-  __ESIMD_NS::simd<float, N> fmod = reminder + abs_x;
-  __ESIMD_NS::simd<float, N> fmod_abs = __ESIMD_NS::abs(fmod);
+  auto fmod_sign_mask = (y.template bit_cast_view<BitMaskType>()) & signBitMask;
+
+  __ESIMD_NS::simd<T, N> reminder =
+      abs_y - abs_x * __ESIMD_NS::trunc<T>(abs_y / abs_x);
+
+  abs_x.merge((T)0.0, reminder >= 0);
+  __ESIMD_NS::simd<T, N> fmod = reminder + abs_x;
+  __ESIMD_NS::simd<T, N> fmod_abs = __ESIMD_NS::abs(fmod);
 
   auto fmod_bits =
-      (fmod_abs.template bit_cast_view<int32_t>()) | fmod_sign_mask;
-  return fmod_bits.template bit_cast_view<float>();
+      (fmod_abs.template bit_cast_view<BitMaskType>()) | fmod_sign_mask;
+  return fmod_bits.template bit_cast_view<T>();
 }
 
 // For Scalar Input
-template <> ESIMD_INLINE float fmod(float y, float x) {
-  return fmod(__ESIMD_NS::simd<float, 1>(y), __ESIMD_NS::simd<float, 1>(x))[0];
+template <typename T> ESIMD_INLINE T fmod(T y, T x) {
+  static_assert(std::is_floating_point<T>::value,
+                "Floating point argument type is expected.");
+  return fmod(__ESIMD_NS::simd<T, 1>(y), __ESIMD_NS::simd<T, 1>(x))[0];
 }
 
 // sin_emu - EU emulation for sin(x)
 // For Vector input
-template <int N>
-ESIMD_INLINE __ESIMD_NS::simd<float, N> sin_emu(__ESIMD_NS::simd<float, N> x) {
-  __ESIMD_NS::simd<float, N> x1;
-  __ESIMD_NS::simd<float, N> x2;
-  __ESIMD_NS::simd<float, N> t3;
-
-  __ESIMD_NS::simd<float, N> sign;
-  __ESIMD_NS::simd<float, N> fTrig;
-  __ESIMD_NS::simd<float, N> TwoPI(6.2831853f);
-  __ESIMD_NS::simd<float, N> CmpI(detail::CMPI);
-  __ESIMD_NS::simd<float, N> OneP(1.f);
-  __ESIMD_NS::simd<float, N> OneN(-1.f);
+template <typename T, int N>
+ESIMD_INLINE __ESIMD_NS::simd<T, N> sin_emu(__ESIMD_NS::simd<T, N> x) {
+  static_assert(std::is_floating_point<T>::value,
+                "Floating point argument type is expected.");
+  __ESIMD_NS::simd<T, N> x1;
+  __ESIMD_NS::simd<T, N> x2;
+  __ESIMD_NS::simd<T, N> t3;
+
+  __ESIMD_NS::simd<T, N> sign;
+  __ESIMD_NS::simd<T, N> fTrig;
+  __ESIMD_NS::simd<T, N> TwoPI((T)detail::__ESIMD_CONST_PI * (T)2.0);
+  __ESIMD_NS::simd<T, N> CmpI((T)detail::__ESIMD_CONST_PI);
+  __ESIMD_NS::simd<T, N> OneP((T)1.0);
+  __ESIMD_NS::simd<T, N> OneN((T)-1.0);
 
   x = esimd::fmod(x, TwoPI);
+  x.merge(TwoPI + x, x < 0);
```

Comment:
Can you please explain this line? I don't see equivalent for it in the code "before" patch.
Does this PR have not only restructuring changes, but also sone math fix?

Reply:
Ok, from phone talk I understood, this algorithm did not work correctly for some negative values.